[Sue Zheng Long] iP

zhenglong1603 opened the pull request:
# Zbot

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://example.com)

Zbot frees your mind of having to remember things you need to do. It's,  

- text-based  
- easy to learn  
- ~~FAST~~ **SUPER FAST** to use  

All you need to do is,  

1. download it from [here](https://example.com).  
2. double-click it.  
3. add your tasks.  
4. let it manage your tasks for you 😊  

And it is **FREE!**  

## Features:  
- [x] Managing tasks  
- [x] Managing deadlines (coming soon)  
- [ ] Reminders (coming soon)  

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:  

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zhenglong1603 committed:
Added greet and exit feature and renamed the bot

zhenglong1603 committed:
added support for echo functionality

zhenglong1603 committed:
Added add task and list tasks functionalities

zhenglong1603 committed:
added task class and taskMarking functionality

zhenglong1603 committed:
Added ToDo, Deadline, and Event task classes and functionality

zhenglong1603 committed:
Added Automated Text UI Testing

zhenglong1603 committed:
Added error handling with exceptions thrown

zhenglong1603 committed:
Added deleteTask functionality, EmptyTaskListException and InvalidTaskNumberException

zhenglong1603 committed:
Add saveFile functionality.

zhenglong1603 committed:
Merge branch 'branch-Level-7'

zhenglong1603 committed:
Add dates and times formatting functionality.

zhenglong1603 committed:
Add ZBOT.txt

zhenglong1603 committed:
Merge branch 'branch-Level-8'

zhenglong1603 committed:
Add OOP

zhenglong1603 committed:
Merge branch 'branch-MoreOOP'

zhenglong1603 committed:
Add packages

zhenglong1603 committed:
Merge branch 'branch-Packages'

zhenglong1603 committed:
Merge branch 'add-gradle-support'

zhenglong1603 committed:
Add Junit testing

zhenglong1603 committed:
Merge branch 'branch-A-JUnit'

zhenglong1603 committed:
Package as JAR FIle

zhenglong1603 committed:
Merge branch 'branch-A-Jar'

zhenglong1603 committed:
Add findTask functionality

zhenglong1603 committed:
Add JavaDoc

zhenglong1603 committed:
Follow Coding Standard

zhenglong1603 committed:
Merge branch 'branch-A-CodingStandard'

zhenglong1603 committed:
Merge conflict resolution for branch-A-JavaDoc with master

Resolved merge conflicts while adding Javadoc comments to task classes.

zhenglong1603 committed:
Merge conflict resolution for branch-Level 9 with master

Resolved merge conflicts while adding Javadoc comments to findTask function

itsvari commented on data/ZBOT.txt:
devops: consider adding this, or the entire `data` directory, to your `.gitignore`

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on gradle/wrapper/gradle-wrapper.jar:
devops: `.jar` and other binaries should probably be added to your `.gitignore`, to keep the repo cleaner. build artefacts should be pushed to your releases page instead!

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,144 @@
+     */
+    public static void parseInput(String input, Ui ui, TaskList taskList) throws IncorrectInputException, InvalidCommandException, EmptyTaskListException, InvalidTaskNumberException {
+        String[] parts = input.split(" ", 2);
+        final String SUPPORTED_COMMANDS = "- list\n- mark\n- unmark\n- find\n- delete\n- todo\n- deadline\n- event\n- bye";
```
this is an interesting way of detecting the commands; is there any reason why you didn't go with an `Enum` instead?

zhenglong1603 replied:
tbh, it just did not cross my mind when i implemented it at first haha. Ill take note in case enum is needed for the subsequent tasks.

itsvari commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,144 @@
+            if (parts.length != 2) {
+                throw new IncorrectInputException("Sorry!! Please ensure your command matches the example: \"delete 1\"");
+            }
+            int deleteIndex = getIndex(parts,taskList);
```
coding standard: there should be a space between the parameters

```suggestion
            int deleteIndex = getIndex(parts, taskList);
```

zhenglong1603 replied:
Thanks for the input! fixed this:)

itsvari commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,144 @@
+     *                                     number of tasks in the task list.
+     */
+    public static int getIndex(String[] parts, TaskList taskList) throws IncorrectInputException, EmptyTaskListException, InvalidTaskNumberException {
+        int markIndex;
```
is there a reason this is called `markIndex`? It's used in more than just the `mark` command; consider renaming to `taskIndex` instead

zhenglong1603 replied:
Thanks for the input! fixed this:)

itsvari commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,145 @@
+            throw new ZBOTFileNotFoundException("The data could not be found");
+        }
+        return ans;
+    }
```
minor nitpick: might be better to use more descriptive variable names in this function. you've got single-letter variable names that refer to different types of objects (a `File`, `Scanner`, `*Task`, and `Exception`). yes this is technically allowed by the coding standard, but having more descriptive names would aid readability in my opinion ^-^

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,145 @@
+                            .append(((EventTask) task).getToDate())
+                            .append("\n");
+                }
+            }
```
there's a lot of duplication happening in the task serialisation logic. might be worth abstracting out to its own portion towards the end of the process.

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,145 @@
+                throw new IOException("Sorry !! Failed to create directory: " + parentDir.getAbsolutePath());
+            }
+        }
+        try {
```
consider making this a try-with-resources, so that the `FileWriter` is automatically closed for you. right now, you're only closing the `FileWriter` within the main `try` block, so if an exception is thrown, you're not actually closing it. this could also be done with a `finally` block.

```suggestion
        try (FileWriter fileWriter = new FileWriter(filePath)) {
```

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/ZBOT.java:
```diff
@@ -0,0 +1,80 @@
+                    System.out.println("---------------------------------------------------");
+                    System.out.println(e.getMessage());
+                    System.out.println("---------------------------------------------------");
+                }
```
consider abstracting this into a function within the `Ui` class! let `Ui` handle all UI-related functionality, and keep this main class relatively lightweight and focused.

zhenglong1603 replied:
abstracted it out! thanks:)

itsvari commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,145 @@
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.List;
```
coding standard: consider cleaning up this list of imports to maintain alphabetical order + logical grouping:

```suggestion
import zbot.exceptions.ZBOTFileNotFoundException;

import zbot.tasks.DeadlineTask;
import zbot.tasks.EventTask;
import zbot.tasks.Task;
import zbot.tasks.TaskList;
import zbot.tasks.ToDoTask;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
```

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/ZBOT.java:
```diff
@@ -0,0 +1,80 @@
+import zbot.tasks.TaskList;
+
+import java.util.Scanner;
+import java.io.IOException;
```
as mentioned above, consider rearranging to alphabetical order

zhenglong1603 replied:
thanks for the input! fixed it:)

itsvari commented on src/main/java/zbot/tasks/Task.java:
consider making this an abstract class! you've already got the three task types, there's probably no reason for a bare `Task` object to be instantiated at this point.

zhenglong1603 replied:
done!

itsvari commented on src/main/java/zbot/Ui.java:
might be worth making a method dedicated to inserting the separating lines, and inserting a message in between! you're duplicating a lot of code to insert lines, and it's spread throughout multiple files. if you ever decide to change the format of ZBOT's responses in the future, you'd only need to change it in this one method, rather than the much larger refactor of finding it in multiple locations throughout your project. it might be manageable at this point with relatively few LoC, but it'll become a much bigger headache at larger project sizes.

zhenglong1603 replied:
Thanks for the input! fixed this:) initially I thought of this problem as well but was still wondering how to go about doing it. Thanks for your suggestion:)

itsvari commented on src/main/java/zbot/tasks/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+        System.out.println("OK, I've marked this task as not done yet:");
+        System.out.println(taskList.get(index).toString());
+        System.out.println("---------------------------------------------------");
+    }
```
as mentioned in the comment on `Ui.java`, consider abstracting the separating line code to its own method in `Ui`, and call that new method everywhere else instead.

zhenglong1603 replied:
thanks for the input! implemented it:)

itsvari commented on src/main/java/zbot/tasks/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+     * @return The list of tasks in the task list.
+     */
+    public List<Task> getTaskList() {
+        return this.taskList;
```
i'm not sure if this is properly preserving encapsulation? is there a reason why you're doing this, instead of returning a copy?

zhenglong1603 replied:
yep fixed this

itsvari commented on src/test/java/zbot/ZBOTTest.java:
sneaky 👉 😎 👉

itsvari commented on src/test/java/zbot/TaskCommandTest.java:
```diff
@@ -0,0 +1,78 @@
+package zbot;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import zbot.tasks.*;
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.List;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
coding standard: consider rearranging import list. have you configured intellij to do the automatic ordering for you yet? if not, [you can follow this guide on how to do so](https://se-education.org/guides/tutorials/intellijCodeStyle.html).

zhenglong1603 replied:
yeap went to research on this and implemented it! didnt know it existed haha

itsvari reviewed (commented):
overall, a pretty neat implementation, and largely adheres to coding standards! also a decent first pass at implementing Gradle and JUnit, though more tests could definitely be added. it's also very well-documented with javadocs, so good job on that (or clever use of an LLM) ^-^

i do have some minor nitpicks and suggestions, but it's nothing major and overall this codebase should be good enough for the rest of the iP!

you should probably include a `.gitignore` file too, because you've pushed some binaries and unnecessary files that should be kept out of the repo for DevOps purposes. the main project has a good starting template, and [GitHub themselves have very good templates for you to use too](https://github.com/github/gitignore)!

zhenglong1603 committed:
Add more OOP, improve JUnit tests, and fix formatting

This commit improves the code by enhancing object-oriented
principles, increasing test coverage, and fixing formatting
inconsistencies.

Changes include:
- Refactoring code to follow better OOP practices
- Adding more JUnit tests to improve reliability
- Fixing formatting for consistency and readability

Let's improve code maintainability and ensure robust testing.

zhenglong1603 committed:
Merge branch 'branch-edits'

zhenglong1603 committed:
Remove data/ZBOT.txt from Git tracking

zhenglong1603 committed:
Remove gradle from Git tracking

zhenglong1603 committed:
Add CheckStyle

zhenglong1603 committed:
Merge branch 'branch-A-CheckStyle'

Junixm commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,169 @@
+        String[] parts = input.split(" ", 2);
+        final String supportedCommands =
+                "- list\n- mark\n- unmark\n- find\n- delete\n- todo\n- deadline\n- event\n- bye";
+        switch (parts[0]) {
+        case "list":
+            if (parts.length == 1) {
+                ui.showContents(taskList);
+            } else {
+                throw new IncorrectInputException(
+                        "Sorry!! Please ensure your command matches the following example and have nothing else. "
+                                + "(e.g. \"list\")");
+            }
```
LGTM

Junixm commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,176 @@
+        StringBuilder taskString = new StringBuilder();
+
+        if (task instanceof ToDoTask) {
+            taskString.append("T | ")
+                    .append(task.getDoneStatus() ? "1 | " : "0 | ")
+                    .append(task.getDescription());
```
Nice indentation

Junixm commented on src/main/java/zbot/Parser.java:
```diff
@@ -0,0 +1,169 @@
+ * task management, file handling, or displaying results.
+ */
+class Parser {
+    /**
+     * This method processes the user's input and triggers the appropriate actions based on the command.
+     * It interacts with the UI and task list to carry out the task.
+     *
+     * @param input the input string entered by the user, containing the command and related arguments
+     * @param ui the UI object used to display messages and interact with the user
+     * @param taskList the list of tasks to be managed and modified based on the input
+     * @throws IncorrectInputException if the input does not follow the correct format
+     * @throws InvalidCommandException if the command entered by the user does not exist or is not recognized
+     * @throws EmptyTaskListException if the task list is empty and a task operation is attempted
+     * @throws InvalidTaskNumberException if the task number is less than 0 or greater
+     *      than or equal to the size of the task list
+     */
```
Nice comments

Junixm commented on src/main/java/zbot/StorageManager.java:
```diff
@@ -0,0 +1,176 @@
+    private static Task getTask(String[] parts) throws IOException {
+        Task task;
+
+        switch (parts[0]) {
+        case "T":
+            task = new ToDoTask(parts[2]);
+            break;
```
LGTM

zhenglong1603 committed:
Add GUI to the project

The project previously had only a CLI-based interface. A GUI enhances usability by
providing a more intuitive way for users to interact with the application.

Let's,
* implement a JavaFX-based GUI framework
* add scene management to support multiple views
* ensure separation of UI and logic for maintainability

zhenglong1603 committed:
Merge branch 'branch-Level-10'

zhenglong1603 committed:
Improve code quality for improved readability and maintainability

Enhanced code readability by reducing repetition and simplifying nested structures.
Refactored arrowhead-style code to improve clarity and maintainability.

These changes help with debugging, optimization, and long-term code quality.

Let's refactor certain parts of the code to
improve code quality.

zhenglong1603 committed:
Add assertions to verify assumptions and ensure functionality

Added assertions to verify key assumptions in the program.
Ensured that expected conditions are met at critical points in the code.

This improves code reliability and helps catch potential issues during development.

zhenglong1603 committed:
Merge pull request #1 from zhenglong1603/branch-CodeQuality

Improve code quality for improved readability and maintainability

zhenglong1603 committed:
Merge branch 'master' into branch-Assertions

zhenglong1603 committed:
Merge pull request #2 from zhenglong1603/branch-Assertions

Branch assertions

zhenglong1603 committed:
Add Assertions to Gradle

zhenglong1603 committed:
Merge branch 'branch-A-Enable-Gradle-Assertions'

zhenglong1603 committed:
Add a new type of command, Undo

Restores the task list to the version before the previous command was executed.

Having this functionality will improve the
effectiveness of this product.

zhenglong1603 committed:
Merge branch 'branch-C-Undo'

zhenglong1603 committed:
Added personality to bot responses

- Enhanced error messages with playful and engaging tone
- Made task completion responses more rewarding and interactive
- Improved user experience by making interactions feel less robotic

This keeps users engaged while maintaining clarity in responses.

zhenglong1603 committed:
Merge branch 'branch-A-Personality'

zhenglong1603 committed:
Fix JavaDoc

Ensure method/class header comments follow the format specified
in the coding standard.

zhenglong1603 committed:
Merge branch 'branch-A-Personality'

zhenglong1603 committed:
Add error handling for input validation

- Add error handling for trailing/leading spaces in commands
to improve input parsing.
- Add error handling when the start date/time is later than
the end date/time to ensure logical consistency.
- Add error handling for non-existent dates (e.g., Feb 30)
to prevent invalid date entries.

These changes are implemented to improve user input validation
and ensure the integrity of task data. They ensure that users receive
clear feedback on incorrect input formats and prevent the creation of
tasks with invalid date ranges or non-existent dates.

zhenglong1603 committed:
Merge branch 'branch-A-MoreErrorHandling'

zhenglong1603 committed:
Add more JUnit tests

zhenglong1603 committed:
Merge branch 'branch-A-MoreTesting'

zhenglong1603 committed:
Enhance GUI

-Enhance the GUI of the chatbot

zhenglong1603 committed:
Merge branch 'branch-A-BetterGui'

zhenglong1603 committed:
Add User Guide

- Add User Guide to provide more information to users on how to use the chatbot
- Add Ui.png as the screenshot of the chatbot
- Set the title of the chatbot

zhenglong1603 committed:
Merge branch 'branch-A-UserGuide'

zhenglong1603 committed:
Add Better GUI

- Change profile pictures
- Remove light blue background

zhenglong1603 committed:
Merge branch 'branch-BetterGui2'

zhenglong1603 committed:
Edit string message for empty list

zhenglong1603 committed:
Edit Ui picture in the user guide

zhenglong1603 committed:
Modify user guide and improve Find Command output for empty results.

zhenglong1603 committed:
Edit Save command to Bye command

zhenglong1603 committed:
Modify user guide for Bye command

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Modify user guide

zhenglong1603 committed:
Fix bug for Undo command to mark and unmark tasks
